### ISSUE TITLE
Device Discovery

### ISSUE THREAD

**Issue by AndrewStuhr** [ottowayi/pycomm3#91]
Hey, hope all is well!

I was recently using pylogix to discover PLC's on my network (https://github.com/dmroeder/pylogix/blob/master/examples/20_discover_devices.py).  That got me wondering whether pycomm3 can do this.  I couldn't find anything, but thought I'd check with you in case I missed it.  If not, have you considered adding device discovery.  Not sure how much work it would take to add, but I think it would be a dope feature to have!

**Comment by ottowayi** [comment]
No there isn't currently any discover functionality.  I'm not opposed to adding, I've just never had a need for it.  I will look into it and see if I can get something working soon.

**Comment by ottowayi** [comment]
I think I have an initial working version, you should be able to do this:

```
from pycomm3 import CIPDriver  #or LogixDriver or SLCDriver

print(CIPDriver.discover())
```

This is in the develop branch, so you can pull from there and test it out if you want.

**Comment by AndrewStuhr** [comment]
I created a virtual environment and installed the devolopment branch.  Here's what I'm getting when running the code.  Any thoughts?

![image](https://user-images.githubusercontent.com/32115408/105249226-a03d6980-5b2c-11eb-8670-afcc83bcabbe.png)

**Comment by ottowayi** [comment]
Sorry I broke some stuff.  I'm working with a friend and doing a lot of internal changes right now and some of that was done in `develop` and not it's own branch.  If you checkout [this commit](https://github.com/ottowayi/pycomm3/tree/b528b164cdcd8ccf855cbd71fc34c291ebc4f2a9)  it should work.

**Comment by AndrewStuhr** [comment]
![livingTribunal](https://user-images.githubusercontent.com/32115408/105251564-a3d2ef80-5b30-11eb-997b-50510ad4c7b7.jpg)

**Comment by AndrewStuhr** [comment]
Sweet, it worked!  

On a separate note, do you have any ideas for how to get the slot number that actually has tags to read from?  After discovering a new device on the network, I'd like to know which slot to use, too.

**Comment by ottowayi** [comment]
Do you mean find which slot number the PLC is in?  Yeah that's possible, I don't have a script ready to go but it should be possible using `discover()` to find the IPs, then use the `get_module_info()` method on each slot to find the one with a device type for PLCs.   RSLinx does more complicated stuff using the Port object to build the tree.  But, you may find it simple enough that if the product name starts with '1756' you know it's a ControlLogix rack which can have a max of 17 slots, so you could just start looping thru them until you find the plc.

**Comment by ottowayi** [comment]
this feature is included in v1.0.0b1, it can be installed using `pip install pycomm3==1.0.0b1`

**Comment by coltman151** [comment]
I have been playing with this a little bit and was having issues getting my Stratix switch to show up. I could see the packet coming back in wireshark.

I removed `and response.raw[12:20] == context:` from _broadcast_discover and was able to see the packet correctly. The sender context coming back from the Stratix is '00000000000'. I'm not sure if this is related, but if you look at the broadcast packet in wireshark, it shows the first part of the sender context under "Max Response Delay". I'm not sure how to verify if the issue is Pycomm or device related. If you have any ideas on how to verify one way or the other, I can try it. 

I apologize for the awful screenshot, but it does include both packets in wireshark, the logs, and the code I modified. I don't have any image processing software on here. I also attached the log, but the logging doesn't return a lot on the discover method.

![Screenshot from 2021-02-22 22-48-02](https://user-images.githubusercontent.com/65498634/108800092-34518500-7560-11eb-8dec-198f101a9fe1.png)
[pycomm3_V1.0.0_Discover.log](https://github.com/ottowayi/pycomm3/files/6026658/pycomm3_V1.0.0_Discover.log)
